File: frontend/src/app/shared/publication-status-text.ts
import { PublicationState, PublicationStatus } from '../models/catalog';

/**
 * Kdy se vlastní záznam propaguje globálně (docs/datovy-model.md, "Uživatelská vrstva nad
 * globálními daty"; prahy v docs/reputace.md) — čistá funkce oddělená od `publication-status.ts`
 * (komponenta), ať se dá otestovat Vitestem bez renderu. Vrací i18n KLÍČ + parametry, ne
 * hotový text — překlad dělá volající komponenta přes `TranslocoService.translate`.
 *
 * "product"/"store" mají u AWAITING_CONFIRMATIONS jiný text (bezkódové zboží najde kdokoli
 * skenem kódu, PENDING obchod nenajde nikdo) — viz docs/reputace.md, "Práh důvěry pro
 * zveřejnění nového záznamu". "observation" nese vlastní text pro PUBLIC/AWAITING, protože
 * cena sama žádný práh nemá, jen dědí stav od blokujícího zboží/obchodu. "edit" (úprava
 * cizího záznamu) je ve výpisu vždy PENDING_MERGE — ostatní stavy tam nikdy nepřijdou.
 */
export type PublicationRecordKind = 'product' | 'store' | 'observation' | 'edit';

export interface PublicationStatusText {
  key: string;
  params?: { received: number; required: number };
}

export function publicationStatusText(
  status: Pick<PublicationStatus, 'state' | 'confirmationsReceived' | 'confirmationsRequired'>,
  kind: PublicationRecordKind,
): PublicationStatusText {
  switch (status.state) {
    case PublicationState.AwaitingConfirmations:
      return {
        key: `my-contributions.status.awaitingConfirmations.${kind}`,
        params: {
          received: status.confirmationsReceived ?? 0,
          required: status.confirmationsRequired ?? 0,
        },
      };
    case PublicationState.HiddenAfterFlags:
      return { key: `my-contributions.status.hiddenAfterFlags.${kind}` };
    case PublicationState.PendingMerge:
      return { key: `my-contributions.status.pendingMerge.${kind}` };
    default:
      return { key: `my-contributions.status.public.${kind}` };
  }
}

File: mobile/app/src/main/java/cz/kvalitacena/ui/contributions/PublicationStatusText.kt
package cz.kvalitacena.ui.contributions

import cz.kvalitacena.R
import cz.kvalitacena.network.PublicationStatus
import cz.kvalitacena.ui.common.UiText

/**
 * Kdy se vlastní záznam propaguje globálně (docs/datovy-model.md, "Uživatelská vrstva nad
 * globálními daty"; prahy v docs/reputace.md) — čistá funkce mimo Compose, ať jde otestovat
 * JUnitem bez instrumentace (stejný vzor jako PriceChartGeometry/StoreLabel). Webový protějšek:
 * frontend shared/publication-status-text.ts.
 *
 * PRODUCT/STORE mají u AWAITING_CONFIRMATIONS jiný text (bezkódové zboží najde kdokoli skenem
 * kódu, PENDING obchod nenajde nikdo) — viz docs/reputace.md. OBSERVATION nese vlastní text pro
 * PUBLIC/AWAITING, protože cena sama žádný práh nemá, jen dědí stav od blokujícího zboží/
 * obchodu. EDIT (úprava cizího záznamu) je ve výpisu vždy PENDING_MERGE.
 */
enum class PublicationRecordKind { PRODUCT, STORE, OBSERVATION, EDIT }

fun publicationStateLabel(state: String): Int = when (state) {
  "AWAITING_CONFIRMATIONS" -> R.string.my_contributions_state_awaiting_confirmations
  "HIDDEN_AFTER_FLAGS" -> R.string.my_contributions_state_hidden_after_flags
  "PENDING_MERGE" -> R.string.my_contributions_state_pending_merge
  else -> R.string.my_contributions_state_public
}

fun publicationStatusText(status: PublicationStatus, kind: PublicationRecordKind): UiText {
  return when (status.state) {
    "AWAITING_CONFIRMATIONS" -> {
      val required = status.confirmationsRequired ?: 0
      val received = status.confirmationsReceived ?: 0
      val resId = when (kind) {
        PublicationRecordKind.PRODUCT -> R.string.my_contributions_status_awaiting_product
        PublicationRecordKind.STORE -> R.string.my_contributions_status_awaiting_store
        PublicationRecordKind.OBSERVATION -> R.string.my_contributions_status_awaiting_observation
        PublicationRecordKind.EDIT -> R.string.my_contributions_status_pending_merge_edit
      }
      UiText.Res(resId, listOf(required, received))
    }
    "HIDDEN_AFTER_FLAGS" -> UiText.Res(
      when (kind) {
        PublicationRecordKind.PRODUCT -> R.string.my_contributions_status_hidden_product
        PublicationRecordKind.STORE -> R.string.my_contributions_status_hidden_store
        PublicationRecordKind.OBSERVATION -> R.string.my_contributions_status_hidden_observation
        PublicationRecordKind.EDIT -> R.string.my_contributions_status_pending_merge_edit
      },
    )
    "PENDING_MERGE" -> UiText.Res(R.string.my_contributions_status_pending_merge_edit)
    else -> UiText.Res(
      when (kind) {
        PublicationRecordKind.PRODUCT -> R.string.my_contributions_status_public_product
        PublicationRecordKind.STORE -> R.string.my_contributions_status_public_store
        PublicationRecordKind.OBSERVATION -> R.string.my_contributions_status_public_observation
        PublicationRecordKind.EDIT -> R.string.my_contributions_status_pending_merge_edit
      },
    )
  }
}
